File: src/2370_longest_ideal_subsequence/longest_ideal_subsequence.py
from typing import List


class Solution:
    def longestIdealString(self, s: str, k: int) -> int:
        dp = [0] * 26
        for char in s:
            curr = ord(char) - ord("a")
            max_len = 0
            for prev in range(26):
                if abs(curr - prev) <= k:
                    max_len = max(max_len, dp[prev])
            dp[curr] = max_len + 1
        return max(dp)


# Test scaffolding
def test():
    sol = Solution()
    print(sol.longestIdealString("acfgbd", 2), "| 4")
    print(sol.longestIdealString("abcd", 3), "| 4")
    print(sol.longestIdealString("abcde", 0), "| 1")
    print(sol.longestIdealString("a", 25), "| 1")
    print(sol.longestIdealString("azazaz", 25), "| 6")


if __name__ == "__main__":
    test()

File: src/2370_longest_ideal_subsequence/longest_ideal_subsequence.cpp
#include <algorithm>
#include <iostream>
#include <string>
#include <vector>
using namespace std;

class Solution {
   public:
    int longestIdealString(string s, int k) {
        vector<int> dp(26, 0);
        for (char ch : s) {
            int curr = ch - 'a';
            int maxLen = 0;
            for (int prev = 0; prev < 26; ++prev) {
                if (abs(curr - prev) <= k) {
                    maxLen = max(maxLen, dp[prev]);
                }
            }
            dp[curr] = maxLen + 1;
        }
        return *max_element(dp.begin(), dp.end());
    }
};

// Test scaffolding
void test() {
    Solution sol;
    cout << sol.longestIdealString("acfgbd", 2) << " | 4\n";
    cout << sol.longestIdealString("abcd", 3) << " | 4\n";
    cout << sol.longestIdealString("abcde", 0) << " | 1\n";
    cout << sol.longestIdealString("a", 25) << " | 1\n";
    cout << sol.longestIdealString("azazaz", 25) << " | 6\n";
}

int main() {
    test();
    return 0;
}
